### PR TITLE
Getting parameters from TransformModule & TransformedDistribution

### PR DESCRIPTION
**Pull request by stefanwebb**
This PR is about discussing a better method for getting parameters out of a transformed distribution.

Previously, you had to do e.g., `x = PlanarFlow(2); x.module.parameters()` to extract the parameters from a single transform. To get the parameters out of a sequence of `Transform`s was tricky because you had to extract the `.module`s from each transform and it was decoupled from `TransformedDistribution`.

I'm suggesting a two-part fix. Firstly, a new class `TransformModule` that turns transforms into true `nn.Module`s, so you can do e.g. `x = PlanarFlow(2); x.parameters()` if `PlanarFlow` inherits from `TransformModule`. Secondly, also making `TransformedDistribution` a true `nn.Module`. Then you can simply pass the parameters from the transformed distribution to e.g. an optimizer by doing `flow_dist = TransformedDistributions(base_dist, flows); optimizer = torch.optim.Adam(flow_dist.parameters())`

### CODE REVIEW

File: pyro/distributions/torch.py [hunk 1]
```diff
@@ -282,11 +282,28 @@ def expand(self, batch_shape):
             return type(self)(df, loc, scale, validate_args=validate_args)
 
 
-class TransformedDistribution(torch.distributions.TransformedDistribution, TorchDistributionMixin):
+class TransformedDistribution(torch.distributions.TransformedDistribution, TorchDistributionMixin, torch.nn.Module):
```
Comment (neerajprad) [inline pyro/distributions/torch.py]: What are the specific methods we would like to inherit from `nn.Module`? I am wondering if it safe to inherit off of `Module` for all `TransformedDistribution` and tying ourselves up in terms of future compatibility. If it is only a couple of methods like `.parameters()` and `.cuda()`, we could provide implementations for these (and use composition when needed to call `.module.cuda()` under the hood, for instance).

File: pyro/distributions/torch.py [hunk 2]
```diff
@@ -282,11 +282,28 @@ def expand(self, batch_shape):
...
+    def __init__(self, *args, **kwargs):
+        torch.distributions.TransformedDistribution.__init__(self, *args, **kwargs)
+        TorchDistributionMixin.__init__(self)
+        torch.nn.Module.__init__(self)
+
+        # Not all transforms have parameters, so only want to extract nn.Module ones
+        self.transform_modules = torch.nn.ModuleList([t for t in self.transforms if isinstance(t, torch.nn.Module)])
```
Comment (neerajprad) [inline pyro/distributions/torch.py]: This will register the parameters of the transform for use in `.parameters` and `.cuda`. Are there cases where we would have to separately grab the base distribution's parameters? If so, it might give a false impression to the user that they have transferred their parameters to the GPU by calling `transformed_dist.cuda()`, but they'll get an error later because their base distribution's params will still be on the CPU.

File: pyro/distributions/torch.py (hunk 1 above)
Comment (stefanwebb) [reply pyro/distributions/torch.py]: I think basically all the main ones, like `apply`, `cpu`, `cuda`, `load_state_dict`, `named_parameters`, `parameters`, `state_dict`, `to`. IMHO, I think future compatibility wouldn't be a problem since `nn.Module` is one of the main abstraction mechanisms in PyTorch and it's been around for so long

File: pyro/distributions/torch.py (hunk 2 above)
Comment (stefanwebb) [reply pyro/distributions/torch.py]: Hmm, at the moment you can't do `.cuda()` or `.parameters()` on a distribution (Fritz explains the rationale [here](https://github.com/pytorch/pytorch/issues/7795#issuecomment-391807661)). I guess the difficulty is created because a transformed distribution mixes objects with and without trainable parameters... I've added in an extra bit of code that adds a base distribution that is also an `nn.Module` to the module list

Comment (stefanwebb) [reply pyro/distributions/torch.py]: I agree it would be confusing if a user calls `.cuda()` and their base distribution in on the `.cpu()`. Although I can't see a good solution. It doesn't seem possible to add a `.cuda()` to distributions in general for the reasons in that issue above. On the other hand, there's a need to have `.cuda()` for normalizing flows, since they have trainable parameters... What about overriding `.cuda()`, `.to()` and `.cpu()` in `TransformedDistribution` to raise an exception? Then the preferred way to transform your normalizing flow to the GPU would be to call `.cuda()` on each `TransformModule` separately (or in a list comprehension)?

Comment (stefanwebb) [reply pyro/distributions/torch.py]: The more I think about this, the more I prefer not making `TransformedDistribution` an `nn.Module`, keeping it lightweight like `Distribution`, and doing all the work with `TransformModule`. I was just looking over the DMM example in Pyro, and I think in a typical use case, you would make an `nn.ModuleList` of `Transforms` inside your model (inheriting from `nn.Module`), and create a `TransformedDistribution` on the fly as needed

Comment (neerajprad) [reply pyro/distributions/torch.py]: I agree - I think that will be a cleaner, more straightforward abstraction!

File: examples/dmm/dmm.py [hunk 3]
```diff
@@ -150,7 +150,7 @@ def __init__(self, input_dim=88, z_dim=100, emission_dim=100,
 
         # if we're using normalizing flows, instantiate those too
         self.iafs = [InverseAutoregressiveFlow(AutoRegressiveNN(z_dim, [iaf_dim])) for _ in range(num_iafs)]
-        self.iafs_modules = nn.ModuleList([iaf.module for iaf in self.iafs])
+        self.iafs_modules = nn.ModuleList([self.iafs])
```
Comment (neerajprad) [inline examples/dmm/dmm.py]: Remove the square brackets, since `self.iafs` is already a list.

File: pyro/distributions/torch.py [hunk 4]
```diff
@@ -287,6 +287,15 @@ def expand(self, batch_shape):
         return super(TransformedDistribution, self).expand(batch_shape)
 
 
+class TransformModule(torch.distributions.Transform, torch.nn.Module):
+    def __init__(self):
+        torch.distributions.Transform.__init__(self)
+        torch.nn.Module.__init__(self)
+
+    def __hash__(self):
+        return super(torch.nn.Module, self).__hash__()
```
Comment (neerajprad) [inline pyro/distributions/torch.py]: nit: I think it will be simpler to just use `hash(self)`.

Comment (stefanwebb) [reply pyro/distributions/torch.py]: This won't work because you need to call the `.__hash__()` from the parent class. Doing `.hash(self)` enters an infinite recursion. Is there a neater way to call the parent class method?

File: examples/dmm/dmm.py (hunk 3 above)
Comment (stefanwebb) [reply examples/dmm/dmm.py]: Ah sorry, yes of course! :)

File: pyro/distributions/torch.py (hunk 4 above)
Comment (neerajprad) [reply pyro/distributions/torch.py]: Sorry, what you have is correct! I was confused - hash(self) would just call `__hash__` method (as it should, I thought it would use `object.__hash__`) itself and go into infinite recursion.

File: pyro/distributions/torch.py [hunk 5]
```diff
@@ -324,6 +333,7 @@ def expand(self, batch_shape):
 
     __all__.append(_name)
 
+__all__.append('TransformModule')
```
Comment (neerajprad) [inline pyro/distributions/torch.py]: I think this can be removed since you are adding this in `distributions.__init__.py`. 

nit: I think a better place for this class might be `torch_distribution.py` which has other mixin classes too.

File: pyro/distributions/torch.py (hunk 1 above)
Comment (fritzo) [reply pyro/distributions/torch.py]: We may need to avoid using `@lazy_property`, or add something to reset all lazy properties when calling `.cpu()`, `.cuda()`, and `.to()`.

File: pyro/distributions/torch.py [hunk 6]
```diff
@@ -287,6 +287,15 @@ def expand(self, batch_shape):
         return super(TransformedDistribution, self).expand(batch_shape)
 
 
+class TransformModule(torch.distributions.Transform, torch.nn.Module):
+    def __init__(self):
+        torch.distributions.Transform.__init__(self)
```
[the remark this answers is not in the document]
Comment (stefanwebb) [reply pyro/distributions/torch.py]: Sounds good! I've made all these changes

**Comment by stefanwebb** [comment]
I also added in docs for `TransformModule`

File: pyro/distributions/torch_patch.py [hunk 7]
```diff
@@ -45,6 +45,19 @@ def _torch_dirichlet_grad(x, concentration, total):
     return unpatched_fn(x, concentration, total)
 
 
+@_patch('torch.distributions.transforms.Transform.__init__')
```
Comment (fritzo) [inline pyro/distributions/torch_patch.py]: Could you just add a comment
```py
# this can be removed when super(...).__init__() is added upstream
```

Comment (stefanwebb) [reply pyro/distributions/torch_patch.py]: Done!
